### PR TITLE
Adjustment to support SCALE compiler which pretends to be NVCC and CLANG at the same time

### PR DESCRIPTION
**Pull request by rbourgeois33**
I am trying to compile Kokkos with the [SCALE compiler](https://docs.scale-lang.com/stable/) to compare backends.

Scale's `nvcc` is a drop in replacement for `nvcc` but is a clang fork. Therefore, when using it, both `__NVCC__` and `__clang__` are true at the same time and leads to a failure of the compile time tests.

This commit fixes this corner case and should not hinder other compile paths

### Related issues / PRs

Kokkos integration in scale's validation test base: https://github.com/spectral-compute/scale-validation/pull/12

### CODE REVIEW

**Review by masterleinad — APPROVED** [review]

File: core/unit_test/TestTypeInfo.cpp [hunk 1]
```diff
@@ -25,7 +25,7 @@ using Lambda = decltype(func);
 #if defined(__NVCC__) && !defined(__CUDA_ARCH__)
 // can't do much
 // it looks like that there is 1st an EDG pass and then a host pass and they cannot both agree on what the type info is
-#elif defined(__EDG__) || (defined(__NVCC__) && defined(__CUDA_ARCH__))
+#elif defined(__EDG__) || (defined(__NVCC__) && !defined(__clang__) && defined(__CUDA_ARCH__))
```
Comment (JBludau) [inline core/unit_test/TestTypeInfo.cpp:29]: Since this is a deviation from our usual guards, could you add an explanatory comment? Otherwise this might just raise eyebrows

Comment (rbourgeois33) [reply core/unit_test/TestTypeInfo.cpp:29]: Would 

`// We check that the compiler is NVCC and not clang, because SCALE's compiler (from spectral compute) is both and should fall into #elif defined(__clang__) `

Be okay ?

Comment (JBludau) [reply core/unit_test/TestTypeInfo.cpp:29]: sure. Maybe add that this is the only place that we have seen problems with this so far, otherwise the question arises why we do not do that everywhere

Comment (science-enthusiast) [reply core/unit_test/TestTypeInfo.cpp:29]: Consider the following comment:
`// We check that the compiler is Nvidia's NVCC. SCALE's compiler (from Spectral Compute) is true for both defined(__clang__) and defined(__NVCC__)`
Also, @JBludau's point about this being the only place with this problem needs to be mentioned.

Comment (rbourgeois33) [reply core/unit_test/TestTypeInfo.cpp:29]: Thanks ! Done

**Review by tcclevenger — APPROVED** [review]
